spring-integration-ip/src/main/java/org/springframework/integration/ip/config/TcpConnectionFactoryFactoryBean.java

```diff
@@ -128,8 +137,11 @@ public final void setBeanFactory(BeanFactory beanFactory) {
 
 	@Override
 	public Class<?> getObjectType() {
-		return this.connectionFactory != null ? this.connectionFactory.getClass()
-											  : AbstractConnectionFactory.class;
+		return this.connectionFactory != null ? this.connectionFactory.getClass() :
+			this.type == null ? AbstractConnectionFactory.class :
+				isServer() ? AbstractServerConnectionFactory.class :
+					isClient() ? AbstractClientConnectionFactory.class :
+						AbstractConnectionFactory.class;
```

Comment:
Got it! (for XML part)
For the ctor concern:
>  because the constructors are different

I don't suggest to instantiate the target object based on the provided class. We still can use internally `client/server` flag based on the `TcpServerConnectionFactory.class.isAssignableFrom(this.type)` condition.

I can merge as is. Just let me know!